eng/scripts/Generate-Patch.ps1

```diff
@@ -0,0 +1,240 @@
+TestPathThrow -Path $SdkDirPath
+TestPathThrow -Path $SetVersionFilePath
+TestPathThrow -Path $UpdateVersionFilePath
+
+. (Join-Path $EngCommonScriptsDir common.ps1)
+
+function GetPatchVersion($ReleaseVersion) {
+  $REGEX_VERSION = '([0-9]+).([0-9]+).([0-9]+)'
+  if(($ReleaseVersion -match $REGEX_VERSION) -and ($Matches.Count -eq 4)) {
+    $MajorVersion = $Matches[1]
+    $MinorVersion = $Matches[2]
+    $PatchVersion = [int]$Matches[3] + 1
+    $PatchVersion = "$MajorVersion.$MinorVersion.$PatchVersion"
+    return $PatchVersion
+  }
+}
+
+function GetRemoteName($MainRemoteUrl) {
+  foreach($Remote in git remote show) {
+    $RemoteUrl = git remote get-url $Remote
+    if($RemoteUrl -eq $MainRemoteUrl) {
+      return $Remote
+    }
+  }
+  return $null
+}
+
+function UpdateChangeLog($ArtifactName, $ServiceDirectoryName, $Version) {
+  
+  $pkgProperties = Get-PkgProperties -PackageName $ArtifactName -ServiceDirectory $ServiceDirectoryName
+  $ChangelogPath = $pkgProperties.ChangeLogPath
+  
+  if (!(Test-Path $ChangelogPath)) {
+    LogError "Changelog path [$ChangelogPath] is invalid."
+    exit 1
+  }
+  
+  $ReleaseStatus = "$(Get-Date -Format $CHANGELOG_DATE_FORMAT)"
+  $ReleaseStatus = "($ReleaseStatus)"
+  $ChangeLogEntries = Get-ChangeLogEntries -ChangeLogLocation $ChangelogPath
+  LogDebug "Adding new ChangeLog entry for Version [$Version]"
+  $Content = @()
+  $Content += ""
+  $Content += "### Dependency Updates"
+  $Content += ""
+  $Content += "Upgraded ``azure-core`` and other dependencies for the library."
+  $Content += ""
+  $newChangeLogEntry = New-ChangeLogEntry -Version $Version -Status $ReleaseStatus -Content $Content
+  if ($newChangeLogEntry) {
+    $ChangeLogEntries.Insert(0, $Version, $newChangeLogEntry)
+  }
+  else {
+    LogError "Failed to create new changelog entry"
+    exit 1
+    }
+  Set-ChangeLogContent -ChangeLogLocation $ChangelogPath -ChangeLogEntries $ChangeLogEntries
+}
+
+if('' -eq $PatchVersion) {
+  $PatchVersion = GetPatchVersion -ReleaseVersion $ReleaseVersion
+  if('' -eq $PatchVersion) {
+    LogError "Could not fetch the patch version. Exitting ..."
+    exit
+  }
+}
+Write-Information "PatchVersion is: $PatchVersion"
+
+$RemoteName = GetRemoteName -MainRemoteUrl $MainRemoteUrl
+if($null -eq $RemoteName) {
+    LogError "Could not fetch the remote name for the URL $MainRemoteUrl Exitting ..."
+    exit
+}
+Write-Information "RemoteName is: $RemoteName"
+
+if('' -eq $BranchName) {
+  $ArtifactNameToLower = $ArtifactName.ToLower()
+  $BranchName = "release/$ArtifactNameToLower/$PatchVersion"
+}
+Write-Information "BranchName is: $BranchName"
+
+try {
+  Write-Information "Reseting the $ArtifactName sources to the release $ReleaseTag."
+  ## Creating a new branch
+  $CmdOutput = git checkout -b $BranchName $RemoteName/main
+  if($LASTEXITCODE -ne 0) {
+    LogError "Could not checkout branch $BranchName, please check if it already exists and delete as necessary. Exitting..."
+    exit
```

Comment:
For all the error cases we should exit with non-zero code so this script will correctly fail in our devops pipelines.